doc/man/gnunet-update.1
.TH GNUNET-UPDATE "1" "31 Jul 2004" "GNUnet"
.SH NAME
gnunet-update \- a tool to update GNUnet
.SH SYNOPSIS
.B gnunet\-update
[\fIOPTIONS\fR] 
.SH DESCRIPTION
.PP
gnunet-update updates gnunets databases after software updates
that change the database format.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (default: /etc/gnunet.conf)
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR \fB\-\-verbose\fR
Be verbose.
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.SH NOTES
gnunet-update was introduced in version 0.6.3.  Updates from
much earlier versions may not be handled properly by the tool.
Always consult the UPDATING file in the GNUnet distribution
for the necessary update procedure.
.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunetd\fP(1), \fBgnunet\-peer\-info\fP(1), \fBgnunet\-transport\-check\fP(1), \fBgnunet\-check\fP(1), \fBgnunet.conf\fP(5)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/server/gnunet-update.c
/*
     This file is part of GNUnet.
     (C) 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunet-update.c
 * @brief tool to process changes due to version updates
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_util.h"

/**
 * Print a list of the options we offer.
 */
static void printhelp() {
  static Help help[] = {
    HELP_CONFIG,
    { 'g', "get", "SECTION:ENTRY",
      gettext_noop("print a value from the configuration file to stdout") },
    HELP_HELP,
    HELP_LOGLEVEL,
    { 'u', "user", NULL,
      gettext_noop("run in user mode (for getting user configuration values)") },
    HELP_VERSION,
    HELP_VERBOSE,
    HELP_END,
  };
  formatHelp("gnunet-update [OPTIONS]",
	     _("Updates GNUnet datastructures after version change."),
	     help);
}

static int be_verbose = NO;

/**
 * Perform option parsing from the command line. 
 */
static int parseCommandLine(int argc, 
			    char * argv[]) {
  int c;
  int user = NO;
  int get = NO;

  /* set the 'magic' code that indicates that
     this process is 'gnunetd' (and not any of
     the user-tools).  Needed such that we use
     the right configuration file... */
  FREENONNULL(setConfigurationString("GNUNETD",
				     "_MAGIC_",
				     "YES"));
  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "get", 1, 0, 'g' },
      { "user", 0, 0, 'u' },
      { "verbose", 0, 0, 'V' },
      { 0,0,0,0 }
    };
    
    c = GNgetopt_long(argc,
		      argv, 
		      "vhdc:g:VL:", 
		      long_options, 
		      &option_index);    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;    
    switch(c) {
    case 'g':
      FREENONNULL(setConfigurationString("GNUNET-UPDATE",
					 "GET",
					 GNoptarg));
      get = YES;
     break;
    case 'L':
      FREENONNULL(setConfigurationString("GNUNETD",
					 "LOGLEVEL",
					 GNoptarg));
     break;
    case 'h': 
      printhelp(); 
      return SYSERR;
    case 'u':
      FREENONNULL(setConfigurationString("GNUNETD",
					 "_MAGIC_",
					 "NO"));
      user = YES;
      break;
    case 'v': 
      printf("GNUnet v%s, gnunet-update 0.0.1\n",
	     VERSION);
      return SYSERR;
    case 'V':
      be_verbose = YES;
      break;
    default:
      printf(_("Use --help to get a list of options.\n"));
      return SYSERR;
    } /* end of parsing commandline */
  }
  if (user && (! get)) {
    printf(_("Option '%s' makes no sense without option '%s'."),
	   "-u", "-g");
    return SYSERR;
  }
  if (GNoptind < argc) {
    printf(_("Invalid arguments: "));
    while (GNoptind < argc)
      printf("%s ", argv[GNoptind++]);
    printf(_("\nExiting.\n"));
    return SYSERR;
  }
  if (get == NO) {
    /* if we do not run in 'get' mode,
       make sure we send error messages 
       to the console... */
    FREENONNULL(setConfigurationString("GNUNETD",
				       "LOGFILE",
				       NULL));
  }
  return OK;
}

static void rename062bCallback(const char * filename,
			       const char * dirName,
			       void * unused) {
  char * oldName;
  char * newName;
  EncName enc;
  HashCode160 hc;
  char * backup;

  if (strlen(filename) < sizeof(HexName)-1)
    return;
  backup = STRDUP(&filename[sizeof(HexName)-1]);
  oldName = MALLOC(strlen(dirName) + strlen(filename) + 10);
  strcpy(oldName, dirName);
  strcat(oldName, "/");
  strcat(oldName, filename);

  if (strlen(filename) != sizeof(HexName)-1) {
    FREE(backup);
    FREE(oldName);
    return;
  }
  if (SYSERR == tryhex2hash(filename,
			    &hc)) {  
    FREE(backup);
    FREE(oldName);
    return;
  }
  hash2enc(&hc, &enc);

  newName = MALLOC(strlen(dirName) + strlen(filename) + strlen(backup) + 10);
  strcpy(newName, dirName);
  strcat(newName, "/");
  strcat(newName, (char*) &enc);
  strcat(newName, backup);
  FREE(backup);
  if (YES == be_verbose)
    printf(_("Renaming file '%s' to '%s'\n"),
	     oldName, newName);
  if (0 != rename(oldName, newName))
    LOG(LOG_ERROR,
	_("Could not rename '%s' to '%s': %s\n"),
	oldName, newName, STRERROR(errno));
  FREE(oldName);
  FREE(newName);
}

#define TRUSTDIR "data/credit/"

/**
 * Update from version 0.6.2b and earlier to
 * 0.6.3 (and later).
 * go over trust/ and hosts/ directories and rename files
 */
static void update062b() {
  char * gnHome;
  char * trustDirectory; 
  
  gnHome = getFileName("",
		       "GNUNETD_HOME",
		       _("Configuration file must specify a "
			 "directory for GNUnet to store "
			 "per-peer data under %s%s\n"));
  trustDirectory = MALLOC(strlen(gnHome) + 
			  strlen(TRUSTDIR)+2);
  strcpy(trustDirectory, gnHome);
  FREE(gnHome);
  strcat(trustDirectory, "/");
  strcat(trustDirectory, TRUSTDIR);
  scanDirectory(trustDirectory,
		&rename062bCallback,
		NULL);
  FREE(trustDirectory);
  gnHome = getFileName("GNUNETD",
		       "HOSTS",
		       _("Configuration file must specify directory for "
			 "network identities in section %s under %s.\n"));
  scanDirectory(gnHome,
		&rename062bCallback,
		NULL);  
}

static int work() {
  int * sbit;
  int version;
  int val;
  
  sbit = NULL;
  if (sizeof(int) == stateReadContent("GNUNET-VERSION",
				      (void**)&sbit)) {
    version = *sbit;
    FREE(sbit);
    switch (ntohl(version)) {
    case 0x0630: 
      printf(_("State is current, no update required.\n"));
      break;
    default:
      printf(_("Unknown version, are you down-grading?\n"));
    }
  } else {
    printf(_("Updating from version pre 0.6.3 (or first run)\n"));
    printf(_("You may also want to run gnunet-check -u.\n"));
    update062b();
    FREENONNULL(sbit);
  }
  val = htonl(0x0630);  
  stateWriteContent("GNUNET-VERSION",
		    sizeof(int),
		    &val);
  return OK;
}



int main(int argc, 
	 char * argv[]) {
  char * get;

  if (SYSERR == initUtil(argc, argv, &parseCommandLine))    
    return 0;
  get = getConfigurationString("GNUNET-UPDATE",
			       "GET");
  if (get != NULL) {
    char * sec;
    char * ent;
    char * val;
    
    sec = get;
    ent = get;
    while ( ( (*ent) != ':') &&
	    ( (*ent) != '\0') )
      ent++;
    if (*ent == ':') {
      *ent = '\0';
      ent++;
    }
    val = getConfigurationString(sec, ent);
    if (val == NULL) 
      printf("%u\n",
	     getConfigurationInt(sec, ent));
    else {
      printf("%s\n",
	     val);
      FREE(val);
    }    
    FREE(get);
    doneUtil();
    return 0;
  }
  work();  
  doneUtil();
  return 0;
}

/* end of gnunet-update */
